lib/src/optimism/composition.rs

```diff
@@ -0,0 +1,296 @@
+                            let compose_journal = to_vec(&prior_output)
+                                .expect("Failed to encode prior aggregation journal");
+                            env::verify(Digest::from(self.compose_image_id), &compose_journal)
+                                .expect("Failed to validate prior aggregation");
+                        }
+                        // Validate context
+                        assert_eq!(self.derive_image_id, prior_output.derive_image_id);
+                        assert_eq!(self.compose_image_id, prior_output.compose_image_id);
+                        assert_eq!(
+                            self.eth_chain_merkle_root,
+                            prior_output.eth_chain_merkle_root
+                        );
+                        // Only append merkle range from preparation outputs
+                        let ComposeOutputOperation::PREP = prior_output.operation else {
+                            bail!("Unsupported! Expected ComposeOutput::PREP")
+                        };
+
+                        // Root of input mountain range should equal prior prep's root
+                        assert_eq!(
+                            prior_range
+                                .root(None)
+                                .expect("Empty mountain range used as input"),
+                            self.eth_chain_merkle_root
+                        );
+
+                        (Some(prior_output.eth_chain_tail_block), prior_range)
+                    } else {
+                        Default::default()
+                    };
+                // Insert chain of blocks into mountain range
+                for block in eth_blocks {
+                    // Validate parent relationship
+                    if let Some((_, parent_hash)) = eth_tail {
+                        assert_eq!(block.parent_hash, parent_hash);
+                    }
+                    // Derive block's keccak hash
+                    let block_hash = block.hash();
+                    // Insert hash into mountain range
+                    mountain_range.append_leaf(block_hash.0, None);
+                    // Mark block as new tail
+                    eth_tail.replace((block.number, block_hash));
+                }
+
+                ComposeOutput {
+                    derive_image_id: self.derive_image_id,
+                    compose_image_id: self.compose_image_id,
+                    operation: ComposeOutputOperation::PREP,
+                    eth_chain_tail_block: eth_tail.expect("No blocks used for preparation"),
+                    eth_chain_merkle_root: mountain_range.root(None).expect("Created empty range"),
+                }
+            }
+            ComposeInputOperation::LIFT {
+                derivation: derive_output,
+                eth_tail_proof,
+            } => {
+                #[cfg(target_os = "zkvm")]
```

Comment:
there's no need to do that on the host. the derive output would have already been validated earlier in the execution before composition starts.

Reply:
I know that it should have been. I was more thinking like a consistency or sanity check especially considering that our tests only run locally.
In that context it would be nice to have a wrapper around `env::verify` that calls env:verify on the guest and validates the input on the host instead.

Reply:
I see what you mean, can we open an issue for that though? I'd rather land this monster without further complexity as that verification will probably require a few changes across multiple files